luavm/src/lib/lua_values/mod.rs

```diff
@@ -1,45 +1,52 @@
+mod lua_float;
+mod lua_int;
+mod lua_obj;
 mod lua_table;
 mod tagging;
 
-use self::{lua_table::LuaTable, tagging::*};
+use self::{lua_float::LuaFloat, lua_int::LuaInt, lua_obj::*, lua_table::LuaTable, tagging::*};
 use errors::LuaError;
 use gc::{Finalize, Gc, Trace};
-use std::mem::swap;
+use std::mem::{swap, transmute};
 
 /// Represents a value in Lua.
 #[derive(Debug)]
 pub struct LuaVal {
-    encoded_ptr: usize,
+    val: usize,
 }
 
 impl Finalize for LuaVal {}
 unsafe impl Trace for LuaVal {
     custom_trace!(this, {
         // only tables have a Gc inside, so there is no need to mark anything else
         if let LuaValKind::TABLE = this.kind() {
-            mark(unsafe { &*table_ptr(this.encoded_ptr) });
+            mark(unsafe { &*table_ptr(this.val) });
         }
     });
 }
 
 impl LuaVal {
     /// Create an empty LuaVal which is equivalent to Nil.
     pub fn new() -> LuaVal {
-        LuaVal {
-            encoded_ptr: LuaValKind::NIL as usize,
-        }
+        LuaVal { val: 0 }
     }
 
     /// Returns the type of the value store in the pointer.
     fn kind(&self) -> LuaValKind {
-        LuaValKind::from(self.encoded_ptr)
+        LuaValKind::from(self.val)
+    }
+
+    /// Interprets the value as a pointer to a LuaObj, and returns a pointer to it.
+    fn boxed(&self) -> *mut Box<LuaObj> {
```

Comment:
I might be tempted to call this `as_boxed` to make clear that no boxing is going on.